src/Microsoft.OData.Client/ALinq/GroupByAnalyzer.cs

```diff
@@ -0,0 +1,529 @@
+            // We could circumvent this by defaulting to [aggregate method]+[aggregate property], e.g., AverageAmount
+            // but this is deferred to a future release
+            if (resultSelector.Body.NodeType != ExpressionType.MemberInit && resultSelector.Body.NodeType != ExpressionType.New)
+            {
+                return false;
+            }
+
+            // Analyze key selector
+            if (!AnalyzeGroupByKeySelector(source, keySelector))
+            {
+                return false;
+            }
+
+            // Analyze result selector
+            GroupByResultSelectorAnalyzer.Analyze(source, resultSelector);
+
+            resourceExpr.Projection = new ProjectionQueryOptionExpression(resultSelector.Body.Type, resultSelector, new List<string>());
+
+            return true;
+        }
+
+        /// <summary>
+        /// Analyzes a GroupBy key selector for property/ies that the input sequence is grouped by.
+        /// </summary>
+        /// <param name="input">The input expression.</param>
+        /// <param name="keySelector">Key selector expression to analyze.</param>
+        /// <returns>true if analyzed successfully; false otherwise</returns>
+        private static bool AnalyzeGroupByKeySelector(QueryableResourceExpression input, LambdaExpression keySelector)
+        {
+            Debug.Assert(input != null, $"{nameof(input)} != null");
+            Debug.Assert(keySelector != null, $"{nameof(keySelector)} != null");
+
+            EnsureApplyInitialized(input);
+            Debug.Assert(input.Apply != null, $"{nameof(input)}.Apply != null");
+
+            // Scenario 1: GroupBy(d1 => [Constant])
+            ConstantExpression constExpr = keySelector.Body as ConstantExpression;
+            if (constExpr != null)
+            {
+                input.Apply.KeySelectorMap.Add(constExpr.Value.ToString(), constExpr);
+
+                return true;
+            }
+
+            // Scenario 2: GroupBy(d1 => d1.Property) and GroupBy(d1 = d1.NavProperty...Property)
+            if (keySelector.Body.NodeType == ExpressionType.MemberAccess)
+            {
+                MemberExpression memberExpr = (MemberExpression)keySelector.Body;
+
+                // Validate grouping expression
+                ResourceBinder.ValidationRules.ValidateGroupingExpression(memberExpr);
+
+                Expression boundExpression;
+                if (!TryBindToInput(input, keySelector, out boundExpression))
+                {
+                    return false;
+                }
+
+                input.Apply.GroupingExpressions.Add(boundExpression);
+                input.Apply.KeySelectorMap.Add(memberExpr.Member.Name, memberExpr);
+
+                return true;
+            }
+
+            // Only proceed beyond here if key selector body is a simple instantiation.

```

Comment:
does this statement mean scenario 4 and 4 below are not supported?

Reply:
From the tests I see they are supported.

Reply:
how will something that is not a simple instantiation look like? curious

Reply:
@ElizabethOkerio This statement states that we will only proceed to scenarios 3 or 4 if we encounter a simple instantiation. The `if` statement does that check. There's no suggestion that scenarios 3 and 4 are not supported.
From the line below this comment, the simple instantiation I'm referring are either member initialization - `ExpressionType.MemberInit` (using initializer list) or constructor initialization - `ExpressionType.New`.

Then, 4 or 5 lines below, I have provided examples of expressions that result into scenarios 3 and 4

Scenario 3: `GroupBy(d1 => new { d1.Property1, ..., d1.PropertyN })` will result into a constructor instantiation based on anonymous type semantics

Scenario 4: `GroupBy(d1 => new Cls { ClsProperty1 = d1.Property1, ..., ClsPropertyN = d1.PropertyN })` - where `Cls` is a declared class - will result into member initialization.

Reply:
I'll reword that statement to: `Check whether the key selector body is a simple instantiation`.